ihmc-java-toolkit/src/main/java/us/ihmc/tools/thread/LoopingThread.java

```diff
@@ -0,0 +1,193 @@
+   {
+      super(target, name);
+      this.target = target;
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   public LoopingThread(RunnableThatThrows targetThatThrows, ExceptionHandler exceptionHandler, String name)
+   {
+      this(targetThatThrows, exceptionHandler, -1.0, name);
+   }
+
+   public LoopingThread(RunnableThatThrows targetThatThrows, ExceptionHandler exceptionHandler, double loopFrequencyLimit, String name)
+   {
+      super(() -> ExceptionTools.handle(targetThatThrows, exceptionHandler), name);
+      target = () -> ExceptionTools.handle(targetThatThrows, exceptionHandler);
+      limitLoopFrequency(loopFrequencyLimit);
+   }
+
+   /**
+    * Limit the frequency of the execution loop.
+    * To un-limit the loop frequency, pass in a number less than or equal to 0.0.
+    * @param frequencyLimit The limit for the loop frequency.
+    *                       If zero or negative, the loop's frequency is not limited.
+    */
+   public void limitLoopFrequency(double frequencyLimit)
+   {
+      loopPeriodLimit = Conversions.hertzToSeconds(frequencyLimit);
+   }
+
+   /**
+    * Starts executing the {@link #runInLoop()} method in a loop.
+    */
+   @Override
+   public void start()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.LOOP);
+         runCommand.notify();
+      }
+   }
+
+   /**
+    * Signal the thread to execute the {@link #runInLoop()} method once, then pause.
+    */
+   public void runOnce()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.RUN_ONCE);
+         runCommand.notify();
+      }
+   }
+
+   private void initializeIfNeeded()
+   {
+      if (runCommand.compareAndSet(RunCommand.INITIALIZE, RunCommand.PAUSE))
+         super.start();
+   }
+
+   /**
+    * Pauses re-executing the {@link #runInLoop()} method upon its completion.
+    */
+   public void pause()
+   {
+      runCommand.set(RunCommand.PAUSE);
+   }
+
+   /**
+    * Finalizes the execution of {@link #runInLoop()} and stops the thread.
+    * The thread cannot be re-started after calling this method.
+    */
+   public void close()
+   {
+      synchronized (runCommand)
+      {
+         runCommand.set(RunCommand.FINALIZE);
+         runCommand.notify();
+      }
+   }
+
+   /**
+    * Signals the thread to finalize execution of {@link #runInLoop()}
+    * and waits until the thread exits.
+    * Same as calling {@link #close()} then {@link #join()}.
+    */
+   public void blockingClose()
+   {
+      close();
+      try
+      {
+         join();
+      }
+      catch (InterruptedException ignored) {}
+   }
+
+   /**
+    * The method that is executed in a loop.
+    * You may either provide this class a
+    * {@link Runnable}/{@link RunnableThatThrows} to run,
+    * or override this method.
+    */
+   public void runInLoop()
```

Comment:
I think this method name is pretty ambiguous for the public API. If some methods do stuff in the user's thread, the ones that don't should have *Async. Like runOnceAsync()

Reply:
I don't understand this comment

Reply:
Are you suggesting `runInLoopAsync()`? I was trying to keep it close to the `Thread` API (`run()`) so it'd be familiar to the user of this class

Reply:
Having thought more, I'm suggesting first to make this method private.

Reply:
That wouldn't allow extending the method, which is part of my goal with this class.

Reply:
Oh, I think I missed that as a goal of this class. I think you should require the runnableThatThrows be passed in at construction and make it final.

Reply:
Do you still mean to remove the `runInLoop()` method?

Reply:
Yes, in that case it wouldn't be needed?

Reply:
I just realized why it's needed. Without a `runInLoop()` method it's difficult to extend `PausableLoopingThread`. 

E.g. a common way to use `Thread` is like: 
```
public class MyClass extends Thread
{
   private final Param myParam;

   public MyClass(/* parameters */)
   {
      this.myParam = myParam;
      ...
   }

   @Override
   public void run()
   {
      // Do something with myParam
   }
}
```

I thought perhaps it'd be possible to do something like:
```
public class MyClass extends PausableLoopingThread
{
   private final Param myParam;
 
   public MyClass(/* params */)
   {
      super(this::run, /* other params */);
      this.myParam = myParam;
   }

   private void run()
   {
      // Do something with myParam
   }
}
```

BUT I don't have access to `this` before the call to `super()`. The natural way to extend `PausableLoopingThread` would be like `Thread`, to override the method that it runs on a thread. 

Perhaps there's a different hack to going around this, but now I feel that the class should have a `runInLoop()` method which can be extended.